### ISSUE TITLE
Dictionary is unsubscriptable false positive

### ISSUE THREAD

**Issue by umnovI** [pylint-dev/pylint#9340]
### Bug description

```python
Value is unsubscriptable error when using `random.sample()` unless I explicitly convert it to list `list(random.sample())`. Then the error goes away.


Code that gives an error:

import json
import random

a = '{"count":64,"next":"https://pokeapi.co/api/v2/berry/?offset=20&limit=20","previous":null,"results":[{"name":"cheri","url":"https://pokeapi.co/api/v2/berry/1/"},{"name":"chesto","url":"https://pokeapi.co/api/v2/berry/2/"},{"name":"pecha","url":"https://pokeapi.co/api/v2/berry/3/"},{"name":"rawst","url":"https://pokeapi.co/api/v2/berry/4/"},{"name":"aspear","url":"https://pokeapi.co/api/v2/berry/5/"},{"name":"leppa","url":"https://pokeapi.co/api/v2/berry/6/"},{"name":"oran","url":"https://pokeapi.co/api/v2/berry/7/"},{"name":"persim","url":"https://pokeapi.co/api/v2/berry/8/"},{"name":"lum","url":"https://pokeapi.co/api/v2/berry/9/"},{"name":"sitrus","url":"https://pokeapi.co/api/v2/berry/10/"},{"name":"figy","url":"https://pokeapi.co/api/v2/berry/11/"},{"name":"wiki","url":"https://pokeapi.co/api/v2/berry/12/"},{"name":"mago","url":"https://pokeapi.co/api/v2/berry/13/"},{"name":"aguav","url":"https://pokeapi.co/api/v2/berry/14/"},{"name":"iapapa","url":"https://pokeapi.co/api/v2/berry/15/"},{"name":"razz","url":"https://pokeapi.co/api/v2/berry/16/"},{"name":"bluk","url":"https://pokeapi.co/api/v2/berry/17/"},{"name":"nanab","url":"https://pokeapi.co/api/v2/berry/18/"},{"name":"wepear","url":"https://pokeapi.co/api/v2/berry/19/"},{"name":"pinap","url":"https://pokeapi.co/api/v2/berry/20/"}]}'

items: list[dict[str, str]] = random.sample(json.loads(a)["results"], 5)
for item in items:
    b = item["url"]
```

### Command used

```shell
vscode-pylint extension
logs output:

~\miniconda3\envs\pokemon_project\python.exe -m pylint --reports=n --output-format=json --disable=C0114 --disable=C0116 --disable=C0103 --max-line-length=120 --clear-cache-post-run=y --from-stdin mytestfile.py
```
```


### Pylint output

```python
{
        "type": "error",
        "module": "test",
        "obj": "",
        "line": 9,
        "column": 8,
        "endLine": 9,
        "endColumn": 12,
        "path": "test.py",
        "symbol": "unsubscriptable-object",
        "message": "Value 'item' is unsubscriptable",
        "message-id": "E1136"
    }
```

### Expected behavior

no error

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.2
Python 3.12.1 | packaged by conda-forge | (main, Dec 23 2023, 07:53:56) [MSC v.1937 64 bit (AMD64)]
```

### OS / Environment

vscode-pylint v2023.11.13481007

vscode:
Version: 1.85.1 (user setup)
OS: Windows_NT x64 10.0.22631
